feat: x:helper

AirQuick opened the pull request:
This pull request introduces `x:helper`.

This XSpec

```xspec
<x:description stylesheet="sut.xsl"
               query="sut"
               query-at="sut.xqm">
  <x:helper stylesheet="helper.xsl" />
  <x:helper package-name="http://example.org/helper"
            package-version="1.0" />
  <x:helper query="helper"
            query-at="helper.xqm" />
  ...
```

will be compiled into

XSLT:

```xslt
<xsl:stylesheet>
  <!-- the tested stylesheet -->
  <xsl:import href=".../sut.xsl"/>

  <!-- User-provided library module(s) -->
  <xsl:import href=".../helper.xsl"/>
  <xsl:use-package name="http://example.org/helper" package-version="1.0"/>
  ...
```

XQuery:

```xquery
...
(: the tested library module :)
import module "sut" at ".../sut.xqm";

(: User-provided library module(s) :)
import module "helper" at ".../helper.xqm";
...
```

and you can use functions or global variables in `helper.xsl` (or `helper.xqm` or the `http://example.org/helper` package) in XPath expressions in XSpec. For example,

helper.xsl:

```xslt
<xsl:function name="my:verifier" as="xs:boolean">
  <xsl:param name="actual-result" />
  ...perform complex verification on $actual-result and return success/failure...
</xsl:function>
```

XSpec:

```xslt
<x:expect label="Defer verification to a helper function"
          test="my:verifier($x:result)" />
```

See the schema annotation for some more details:

https://github.com/xspec/xspec/blob/978f081328482df3efb80b627dd0aff5d35dd955/src/schemas/xspec.rnc#L43-L49

@galtm (https://github.com/xspec/xspec/issues/1090#issuecomment-662447413)
> Do you think `x:include-helper` should support the import/include distinction, or would that be out of scope?

In this implementation, the helper is not included but imported in order to protect the compiled stylesheet from the helper.
Since the helper is imported after the SUT, the helper has higher precedence than the SUT. It is tested in `test/helper_stylesheet.xspec` by `override.xsl`.

@jcronk 
@mricaud
This is an attempt to implement what I provisionally called `x:include-helper`.

vincentml commented:
This sounds like it could be useful in many situations where the output of the XSLT/XQuery/Schematron needs to be normalized in some way to work in an XSpec test scenario.

Here is an example of an XSpec scenario from a recent project where the proposed x:helper might simplify things. The sanitize-expected.xsl is actually run on the "clean" files that are being compared against, not on the result of the SUT, and it does several normalization steps such as removing default attributes added by parsing against a DTD, and replacing generated id values and timestamps with `...`. Normalizing the expected XML via an XSLT as part of running the XSpec test allows the "clean" expected examples to be pristine examples of the expected result.

```xml
           <x:scenario label="ICNV_A_1695816_J.xml">
                <x:context href="../test/input/ICNV_I_38_01_J/ICNV_A_1695816/ICNV_A_1695816_J.xml"/>
                <x:expect label="complete" select="transform(map{
                    'source-node': doc('../../test/expected/ICNV_I_38_1_J/ICNV_A_1695816/ICNV_A_1695816_J.xml'), 
                    'stylesheet-location': '../sanitize-expected.xsl'})?output/*"/>
            </x:scenario>
```

The sanitize-expected.xsl normalization is used on some, not all, of the scenarios in the XSpec test file.

I think the proposed x:helper could be improved to work for all of these use cases:

* use the helper to normalize the result of the SUT
* use the helper to normalize the expected result
* use the helper in specific scenarios

Could something like the following be added to the implementation to support these use cases?

1. Add a `name` attribute to `x:helper` to give a name to each helper.
2. Add a `use-helper` attribute to `x:context` to declare that a helper should be used to normalize the result of the SUT 
3. Add a `use-helper` attribute to `x:expect` (and related elements) to declare that a helper should be used to normalize the expected result

vincentml commented:
Perhaps the x:helper might already work in the three use cases that I mentioned. If the XSLT or module imported via x:helper contains a normalization function that uses xsl:transform, then the the normalization function could be used in the select attribute of x:context or x:expect.

AirQuick commented:
Thanks @vincentml for sharing your ideas. Yes, I think that if the user writes a normalizer function in a helper, the normalizer could be applied in many places.
We could possibly have `x:helper/@name` + `(x:context|x:expect)/@use-helper`, but satisfying a variety of use cases in that way would require more additional parameters, I'm afraid.

galtm commented on src/schemas/xspec.rnc:
```diff
@@ -37,7 +37,34 @@ description =
+		(helper | global-param | import | scenario | pending | variable)+ }
+
+helper =
+	## The test runner will import this stylesheet (XSLT) or query (XQuery) module after importing
+	## the SUT module.
```
It seems that this feature works with `run-as="import"` (the default) and not `run-as="external"`. Should this description say anything about that? In any case, whenever the wiki documents this feature, I think it's worth mentioning that to set the scope.

AirQuick replied:
`run-as=external` allows helper components to be used in some places. I added a test in 706f61e43f2fdb6a2dbca3f080f273b7e41f9ba1.

To summarize,

- `run-as=import`:
  - Helpers can be used in most places.
  - Helper templates are applied to the context set up by `x:context`.
  - Helper stylesheets can override SUT. You would need to be careful not to interfere with SUT too much.
- `run-as=external`:
   - Helpers do not work in `x:call/(@function|@template)`.
   - Helper templates are not applied to the context.
   - Helpers do not affect SUT.

galtm replied:
Glad to hear the feature is more applicable to the external situation than I had thought at first.

When you adapt your comment above for use in the wiki, can you clarify the point about being careful not to interfere with the SUT "too much"? Would it be clearer for the wiki to say something about ensuring that overrides are suitable and do not have adverse side effects in your test? Maybe that is still vague.

AirQuick replied:
Good point. How about this?

> * Helper stylesheets can override the test target stylesheet.
>   For example, if you want to change the behavior of the test target for the purposes of unit testing, you can write some templates in a helper stylesheet and override the test target stylesheet. That's exactly the same as what you usually do using `xsl:import`.
>   Make sure your helper module fulfills only a specific and limited purpose, otherwise your helper may produce adverse side effects on the behavior of the test target and turn your tests meaningless. Likewise the test target may also have side effects on your helper and you may need to protect the helper against it ([example](https://github.com/xspec/xspec/blob/5b01d612b142afead4e438812d56c46f0c4356f8/tutorial/helper/ws-only-text/test-helper.xsl#L21-L24)).

galtm replied:
I like it. Just a few editorial-type suggestions in the last paragraph:
Make sure your helper module fulfills only a specific and limited purpose. Otherwise, your helper may produce adverse side effects on the behavior of the test target and make your tests misleading. Likewise, the test target...

AirQuick replied:
Thanks for the edit. I took it and pushed https://github.com/xspec/xspec/wiki/Integrating-Your-Own-Test-Helpers

galtm commented on src/compiler/generate-common-tests.xsl:
```diff
@@ -188,7 +188,9 @@
    </xsl:template>
 
-   <xsl:template match="@href" as="attribute(href)" mode="x:gather-specs">
+   <!-- TODO: Perhaps, @query-at hint should not always be resolved??? (e.g. for MarkLogic) -->
```
What is the `@query-at hint`? I don't understand this comment. (I don't use MarkLogic.)

AirQuick replied:
I don't understand it either :)
In 319a5b208c8b274d969c57ab46f41dc7b98fa1fe, I added the source of this comment.

galtm replied:
Thanks, the reference was helpful! I didn't notice that the comment wasn't new. (In the earlier context, the comment a few lines earlier mentions that the query file might be in a database rather than on the file system. Now the TODO comment makes a little more sense to me.)

galtm commented on src/compiler/generate-query-tests.xsl:
```diff
@@ -72,6 +72,12 @@
 
+      <xsl:if test="x:helper">
+         <xsl:text>&#x0A;</xsl:text>
+         <xsl:text>(: User-provided library module(s) :)&#x0A;</xsl:text>
```
Lowercase "user-provided"? Please see my similar comment in generate-xspec-tests.xsl.

AirQuick replied:
Fixed in 52d188fc3ac805d09aec692f9065c1daaa359900

galtm commented on src/compiler/generate-xspec-tests.xsl:
```diff
@@ -49,6 +49,11 @@
             <import href="{$stylesheet-uri}" />
          </xsl:if>
 
+         <xsl:if test="x:helper">
```
Should this `<xsl:if>` block go inside the `<xsl:if test="not($is-external)">` block on line 47, because the feature doesn't operate in tests with `run-as="external"`?

AirQuick replied:
https://github.com/xspec/xspec/pull/1103#discussion_r461533350

galtm commented:
A side benefit of this feature for people who use the XSLT Code Coverage functionality is that the helper file doesn't get included in the coverage report. Without this feature, including or importing a helper file in your SUT causes the coverage report to include the helper (XSpec can't distinguish between "helper" and "SUT"), yet it's not the code whose coverage you really want to measure.

AirQuick commented on src/compiler/generate-xspec-tests.xsl:
```diff
@@ -49,6 +49,11 @@
          </xsl:if>
 
+         <xsl:if test="x:helper">
+            <xsl:comment> User-provided library module(s) </xsl:comment>
```
Fixed in 52d188fc3ac805d09aec692f9065c1daaa359900

AirQuick commented on src/schemas/xspec.rnc:
```diff
@@ -37,7 +37,34 @@ description =
+	## The test runner will import this stylesheet (XSLT) or query (XQuery) module after importing
+	## the SUT module.
+	## Specifying an XSLT package will just insert the equivalent <xsl:use-package> in the test
+	## runner. To get the inserted <xsl:use-package> work, the user should set up a Saxon
```
Fixed in 293179f9b40dcbc46efa682ab1c5581b4210ce60

AirQuick commented:
Thanks for the reivew, @galtm 

My initial test set was a bit too rough. I revised them:

- `run-as=external`: 706f61e43f2fdb6a2dbca3f080f273b7e41f9ba1
- XSLT package: 267d15daae4d780a79cd1f6e94ee52ef7964dc23
- Override: ddfc29d5f526cffef4b7a2fb9eab59327cf44807
- Schematron: 63432a214d7761b64ba7e669a838c78dd1a28f56

mricaud commented:
A bit late, but yes I like this `<x:helper>` strategy. It will be really usefull in many case I think.

In issue #1090 I used to propose a "schematron allowForeign like solution" with xsl statement directly written within xspec, but <x:helper> sounds better: it works with xquery and it's implementation independant (it doesn't require le XSpec compiler to be in XSLT).
About the implementation, it looks fine, althought I couldn't understand erevrything. What is "SUT" by the way ?

A few related comments:

- in #1090 we have also consider being able to interact with the _comparison_ algorithm (@AirQuick suggest a `x:delete-patterns()` utility function). This is another way of dealing with your `sanitize-expected.xsl` @vincentml
- Looks like I'm not the only one to use real case expected result ("pristine examples") in my XSpec, just like @vincentml here or @tofi86 in #227. Other might prefer writing the expected result by hand (writing `...` where `generate-id()` is used for example). But I guess we have to consider both way of doing and  `<x:helper>` helps in this direction !
- I didn't know `xsl:use-package` could be used by importing the file in an `xsl:import` previous statement. I thought the pakage has to be in the classpath. Would a function with `visibility="private"` be accessible? yes with `xsl:import` but no with `xsl:use-package`, what about both?
- By reading this issue I couldn't help thinking about <xsl:import-schema> (SaxonEE only). Would it make sens to validate some chunk of xml generated by an XSLT in XSpec? One could actually import a schema within an imported xsl with `<x:helper>` and then call validation function in x:expect. But this is maybe not be the goal of XSpec, in my case I use a maven plugin to validate external XSpec files samples. But I could not valide the result of a function that way. Just a thought...

AirQuick commented:
> What is "SUT" by the way ?

"System Under Test" or a stylesheet/query specified in `/x:description/(@stylesheet|@query)`, as described in [Under the Hood](https://github.com/xspec/xspec/blob/master/tutorial/under-the-hood/Compilation.md#sut).

> * I didn't know `xsl:use-package` could be used by importing the file in an `xsl:import` previous statement. I thought the pakage has to be in the classpath. Would a function with `visibility="private"` be accessible? yes with `xsl:import` but no with `xsl:use-package`, what about both?

A `x:helper`

```xspec
  <x:helper package-name="http://example.org/helper"
            package-version="1.0" />
```

is simply compiled into

```xslt
  <xsl:use-package name="http://example.org/helper" package-version="1.0"/>
```

so the usual restriction of the XSLT package is still applied:
- The package information needs to be provided (usually by a Saxon configuration file like [this](https://github.com/xspec/xspec/blob/master/test/xslt-package/arith/package/config.xml)), otherwise the compiled stylesheet cannot find the specified package.
- Private functions are not accessible.

AirQuick commented:
In 30e8def07e9e4223b226f74a8f5c0df7bc66080b, I added a set of tutorial files to demonstrate a helper stylesheet. It also demonstrates a possible way to remove whitespace-only text nodes from `@href`. If that is fine, I'll
- merge this pr
- add a section to [Wiki](https://github.com/xspec/xspec/wiki/Writing-Scenarios) and explain `x:helper` (copy and paste from [the schema annotation](https://github.com/xspec/xspec/blob/30e8def07e9e4223b226f74a8f5c0df7bc66080b/src/schemas/xspec.rnc#L42-L49) and [a comment](https://github.com/xspec/xspec/pull/1103#discussion_r461533350))
- update [Wiki "Whitespace-only Text Nodes" section](https://github.com/xspec/xspec/wiki/Writing-Scenarios#whitespace-only-text-nodes) and direct the reader to the added tutorial files

AirQuick commented on test/helper_override.xspec:
```diff
@@ -0,0 +1,15 @@
+		<x:call function="my:square">
+			<x:param select="2" />
+		</x:call>
+		<x:expect label="should return a overridden value" select="1234567890" />
```
Fixed in 5b01d612b142afead4e438812d56c46f0c4356f8

AirQuick commented:
> * merge this pr

Done

> * add a section to [Wiki](https://github.com/xspec/xspec/wiki/Writing-Scenarios) and explain `x:helper` (copy and paste from [the schema annotation](https://github.com/xspec/xspec/blob/30e8def07e9e4223b226f74a8f5c0df7bc66080b/src/schemas/xspec.rnc#L42-L49) and [a comment](https://github.com/xspec/xspec/pull/1103#discussion_r461533350))

[Done](https://github.com/xspec/xspec/wiki/Integrating-Your-Own-Test-Helpers)

> * update [Wiki "Whitespace-only Text Nodes" section](https://github.com/xspec/xspec/wiki/Writing-Scenarios#whitespace-only-text-nodes) and direct the reader to the added tutorial files

[Done](https://github.com/xspec/xspec/wiki/Whitespace-only-Text-Nodes#remove-whitespace-only-text-nodes-after-loading-a-document)

I had to restructure Wiki:

- **Getting Started** section in the Sidebar was too much crowded.
  I created **Advanced Topics** section and moved some topics to there from **Getting Started**.

- **Writing Scenarios** topic was bloated too much.
  I extracted two sections (**Whitespace-only Text Nodes** and **Testing Dynamic Errors**) and put them into their own topics.

> * publish a new version of the Oxygen add-on containing this feature

#1138
